### DISCUSSION TITLE
Next.js Middleware and next.config.js rewrite

### DISCUSSION THREAD

**Discussion by anjar** [vercel/next.js#35231] [Help]
Hi all, are next.js middleware can work with next.js rewrite?, i have problem with using those two.

for example, i have rewrite config in next.config.js like this:

``module.exports = {
  async rewrites() {
    return {
      beforeFiles: [
       { source: "/dasbor", destination: "/dashboard" },
      ],
    }
  },
}``

i have _middleware in /dashboard directory, when i access /dashboard, middleware is working and executed, but when i access /dasbor, it not executed

thank you

**Comment by icyJoseph** [comment]
Hi,

Yeah it looks like from the `next.config.js`, rewrites don't trigger middleware. Perhaps something to fix.

However, I've found myself using a root `_middleware`, that is `pages/_middleware.js`. And handle redirecting from there.

```js
import { NextResponse } from "next/server";

export function middleware(req) {
  const { pathname } = req.nextUrl;

  if (pathname === "/dash") {
    return NextResponse.redirect(new URL("/dashboard", req.nextUrl));
  }

  return NextResponse.next();
}
```

Then you could even write tests to protect this behaviour.

## Caveats

A root level `pages/_middleware` will look at all requests, `api/routes`, `pages`, `favicon.icon`, etc.

You could use:

```js
import { NextResponse } from "next/server";

export function middleware(req) {
  const { pathname } = req.nextUrl;

  if (
    pathname.startsWith("/api") || //  exclude all API routes
    pathname.startsWith("/static") || // exclude static files
    pathname.includes(".") // exclude all files in the public folder
  ) {
    return NextResponse.next();
  }

  if (pathname === "/dash") {
    return NextResponse.redirect(new URL("/dashboard", req.nextUrl));
  }

  return NextResponse.next();
}
```

@shuding Will there ever be support for `_middleware` on `beforeFiles`?

**Comment by keslert** [comment]
Ran into this issue today. I want to a/b test my homepage using middleware to choose which page to show. I don't want to add middleware to /pages/_middleware because my site is massive and I don't want to trigger an edge function for every single page request. Unfortunately, as @icyJoseph pointed out, this doesn't appear to be possible right now.

``` typescript
// next.config.js
async rewrites() {
    return {
      beforeFiles: [
        { source: "/", destination: "/homepage" },
      ],
  }
}
```

```typescript
// /homepage/_middleware.ts
import { NextRequest, NextResponse } from "next/server";

const COOKIE_NAME = "homepage-version";

const MARKETING_BUCKETS = ["og", "a"] as const;
const getBucket = () =>
  MARKETING_BUCKETS[Math.floor(Math.random() * MARKETING_BUCKETS.length)];

export function middleware(req: NextRequest) {
  const bucket = req.cookies[COOKIE_NAME] || getBucket();
  const res = NextResponse.rewrite(`/homepage/${bucket}`);

  if (!req.cookies[COOKIE_NAME]) {
    res.cookie(COOKIE_NAME, bucket)
  }

  return res;
}
```

**Reply by ayazhanZimran** [reply]
Hi @keslert,
I ran into the same issue. Did you find a solution to it? If yes, could you please share it?

**Reply by keslert** [reply]
Yes, Next.js changed how Middleware works, so now it is much easier to handle this use case. The new middleware has a matcher which you can use to target only the homepage. https://nextjs.org/docs/advanced-features/middleware

**Comment by MoritzCooks** [comment]
If you have a rewrite configured and which also matches the middleware matcher, the rewrite configuration for that link won't work.
I had a rewrite configured in next.config to rewrite every request for `/de/.*/favicon.ico` to `/favicon.ico`
```TS
// next.config.ts
  async rewrites() {
    return [
     ...
      {
        source: '/de/:path*/favicon.ico',
        destination: '/favicon.ico',
      },
    ]
  }
```

All routes that get matched by the middleware matcher won't be affected by that rewrite as the execution order is the following:
1. **headers** (next.config.js)
2. **redirects** (next.config.js)
3. **Middleware**
4. **rewrites** (next.config.js)

To solve this you have to exclude the specific path in middleware matcher:
```TS
// middleware.ts
export const config = {
  matcher: ['/de/recipe/((?!favicon.ico))([^\\/]*)'], // exclude calls to `/de/recipe/favicon.ico` so next.config rewrite still works
}
```
